Nested FSM Support

Yogeb-tech opened the pull request:
Hello!

I wanted to get your thoughts on a NestedFSM node I implemented. It derives from State and initializes a self-contained FSM that inherits values from the main FSM. Additionally I've also created a demo similar to the blacksmith demo to showcase this new functionality.

Looking ahead, I'm considering overloading the `_on_enter`, `_on_process`, and `_on_exit` functions to give users more control.
For example, in a fighting game, a "Crouching" nested FSM could use its `_on_enter` function to play the crouching animation. This ensures the animation plays before any sub-state (like "Crouch Jab") runs, eliminating the need to manually trigger it in every possible transition or every sub-state's `on_enter` function into the crouching posture.

Do you have any suggestions or concerns about this direction?

Yogeb-tech committed:
prototype complete

Yogeb-tech committed:
base implementation done

ThePat02 commented:
Yo! I already noticed you working on that. I will take a look later that week, as it is something I have been wanted to add for some time now, but the workaround of nesting a fsm inside a integrated behaviour tree was more comfortable (or less work).

> Looking ahead, I'm considering overloading the _on_enter, _on_process, and _on_exit functions to give users more control.

What's the difference from overloading the state like you would normally? This is what you do with all states anyways, so I do not see why this would be a problem.

Yogeb-tech commented:
I'm very glad to see you approve of this implementation! Hopefully I can get back to work on this now that I have more time on my hands. I plan to implement more robust transition checks for the Transition class. This would include generating warnings for cross-hierarchy transitions, such as when a nested FSM attempts to transition into an ancestor FSM.

Then after some clean up, comments, and a template I feel it should be finished.

Yogeb-tech committed:
transition warnings implemented

Yogeb-tech committed:
improved transition warnings +  added comments to transition warnings

Yogeb-tech committed:
added comments to NestedFSM, next is adjusting configuration warnings in states

Yogeb-tech commented:
I'm just now realizing there is a lot more work than I thought to fully integrate NestedFSMs beyond adding configuration warnings for transitions into this plugin. I'm just going to continue working until I feel it's finished lol.

Yogeb-tech committed:
added configuration warnings to NestedFSM and FSM

The new configuration warnings check if the initial state exists within that FSM, preventing human error.

Yogeb-tech committed:
added nestedFSM to UI + created template + added to docs

Yogeb-tech committed:
Update documentation.md

Yogeb-tech committed:
improved transition configuration warnings. They are more performant now

Yogeb-tech committed:
Merge branch 'main' of https://github.com/Flavore669/BehaviourToolkit

Yogeb-tech committed:
minor change

Yogeb-tech commented:
Ok I think nested FSMs are in a good state.

- Added script templates where appropriate
- Added to the toolbox in the UI
- Added to the documentation
- Added a configuration warning to `FSMTransition`, that emits when a user selects a next transition state that is part of a outside of the FSM/NestedFSM's scope
     - The implementation is a bit shaky. I tried to keep it all self-contained in `FSMTransition`. I would like to see your feedback on it, perhaps I'm a bit paranoid.
- Added a configuration warning to `NestedFSM`, that emits when a user selects an initial state that is part of a different FSM/NestedFSM.
     - This was also added to `FiniteStateMachine`. I feel it's not out of the question people would run state machines in parallel

Any feedback is welcome!

ThePat02 commented:
I will do some testing and experimenting! Thanks for your addition! Looks good so far!

Yogeb-tech committed:
Merge branch 'main' into main

AlexValder commented:
Hi, I have several questions regarding this PR.

1. What are current plans for it? I really like this addon, but not being able to nest two FSMs is sad.
2. I have experimented with this change for a while in my project, and I found out that you can't select NestedFSM for LeafFSMEvent and neither can you fire events in it. Is this by design?
3. Will there be a separate icon for nested FSMs?

ThePat02 commented:
Hi @AlexValder 

You can currently nest FSMs by nesting a Behaviour Tree inside of your FSM and then nesting another FSM inside the BT! Not ideal, but it is a valid work around.

I have been hesitant to merge this, since I did not have much time to review and test the implementation, as my focus has been on other projects and I do not want to introduce any issues by merging new features without taking a thorough look.

As far as I can tell after a quick look, the nested FSM inherits from FSMState and is thus not available to be picked by the Leaf. It is treated as a state, rather than a FSM.

Yogeb-tech commented:
I just wanted to chime in to say when you do get the time to look it over I'm open to fixing any issues you find with it. I'm committed to seeing this feature through and happy to help with anything that comes up along the way.

ThePat02 commented:
@Yogeb-tech Since I do not really use this plugin actively right now, I would be open to tag the implementation as [experimental](https://docs.godotengine.org/en/4.4/tutorials/scripting/gdscript/gdscript_documentation_comments.html#tags) and just merge it in. What do you think about that?

Yogeb-tech commented:
That sounds like a good path forward. Would we just let users know about the branch in the documentation, or would you consider uploading the experimental branch to the asset store for convenience?

ThePat02 commented:
Take a look at the link above. I would mark it as `@experimental` in the docstring and push it to the asset store. We can also add a disclaimer in the docstring / patchnotes as well.

Yogeb-tech commented:
Ah I see. I’m not sure why I was over complicating it when this approach works nicely. Yes this sounds good I'm for it. 

I will add the experimental class tag today then.

Yogeb-tech committed:
added @experimental tag, removed unnecessary comments

ThePat02 commented on addons/behaviour_toolkit/ui/toolkit_ui.tscn:
We should mark it as experimental in the UI as well. Either with an icon + tooltip or a print message when clicking this button. What do you think?

Yogeb-tech replied:
I'm leaning towards a warning icon + tooltip approach. It feels more natural to me for the warning to appear in the UI rather than in the output log. I can work on the tooltip and adding a warning icon next to the main one.

Also I do feel it would be beneficial to give it a custom icon before merging it in. Given my implementation of a state machine within a state, I imagine it could be something like the state's circle icon wrapped around the FSM box icon. 

I'm not sure if it's that serious in the first place, but figured I'd pitch the idea since it seems to be a concern.

ThePat02 commented on addons/behaviour_toolkit/finite_state_machine/nested_fsm.gd:
```diff
@@ -0,0 +1,100 @@
+## @experimental: Ready for use, but still awaiting full testing
```
Change the docstring to the following for a clearer wording.

> As an experimental feature, this may still contain bugs. Please be aware that its API and  functionality are subject to change in future updates.

ThePat02 reviewed (changes requested):
Almost good to go. Take a look at the comments please.

Yogeb-tech committed:
updated docstring for a clearer wording

Yogeb-tech committed:
comment out debug prints in fsm_transition

Yogeb-tech committed:
edited comments in nestedfsm

Yogeb-tech committed:
added tooltip

ThePat02 commented:
Neat. I will update the asset lib version in the following days!
